Add/featured video method

mikeyarce opened the pull request:
Fixes #5607

### Changes proposed in this Pull Request

* Removes `sensei_can_user_view_lesson` from legacy `sensei_lesson_video` method.  I don't think it makes sense in an output function, it should be checked later on.  I don't think this would be a problem because there are likely checks at the template level that should account for this. 
* Creates a new Utility function `get_featured_video_html` which will get either the `sensei-lms/featured-video` HTML or parse the `_lesson_video_embed` postmeta and output an oEmbed for the video. 

Example output of `_lesson_video_embed`: 

```
"<div class="video sensei-video-embed">
<iframe width="525" height="295" src="https://www.youtube.com/embed/dQw4w9WgXcQ?feature=oembed" frameborder="0" allowfullscreen></iframe>
</div>"
```

Example Featured Video Block output for Media files:

```
"<figure class="wp-block-video">
<video controls src="http://sensei.local/wp-content/uploads/2022/09/Snap-Camera-Video-2.mp4"></video>
</figure>"
```

Example Featured Video Block output for oEmbeds:
```
"<figure class="wp-block-embed is-type-video is-provider-youtube wp-block-embed-youtube wp-embed-aspect-16-9 wp-has-aspect-ratio">
<div class="wp-block-embed__wrapper">\n
https://www.youtube.com/watch?v=dQw4w9WgXcQ\n
</div>
</figure>"
```
### Testing instructions

* Create a lesson with a Featured Video block and add a video
* Install and activate the Classic Editor plugin
* Add a Lesson with the Classic Editor 
* Add a YouTube link to the `Video Embed Code:` section
* You can use `Sensei_Utils::get_featured_video_html($POST_ID);` in wp shell to see the output for both the Block and Classic Editor posts

mikeyarce committed:
Remove capability check from frontend output function

It doesn’t make sense to check for capabilities on a video output method.

mikeyarce committed:
Add method to get featured video

Get’s featured video block and has a fallback for legacy video embed postmeta

mikeyarce committed:
Add @since to function

mikeyarce committed:
Fix yoda condition

mikeyarce committed:
Add parameter comment

onubrooks commented on includes/class-sensei-frontend.php:
```diff
@@ -835,7 +835,7 @@ public function sensei_frontend_messages() {
 	 * @param int $post_id Optional. Lesson ID. Default 0.
 	 */
 	public function sensei_lesson_video( $post_id = 0 ) {
-		if ( 0 < intval( $post_id ) && sensei_can_user_view_lesson( $post_id ) ) {
```
I get the point that it seems unnecessary but this is legacy code and it may have been added for a reason back then. Is there any other reason you are removing it, maybe a performance issue or something else?

yscik replied:
Yeah, lesson videos are only viewable for enrolled users. We can break out the code getting the video from the meta into a new method and use that in `get_featured_video_html`, and keep this one for the hook. That new method can also return the HTML instead of echoing, so the output buffering wrapper can be left out.

mikeyarce replied:
Thanks @onubrooks and @yscik - I will address this!

onubrooks commented on includes/class-sensei-utils.php:
```diff
@@ -2626,6 +2626,31 @@ public static function format_last_activity_date( string $date ) {
+	 *
+	 * @return string The featured video HTML output
+	 */
+	public static function get_featured_video_html( $post_id ) {
+		if ( has_blocks( $post_id ) ) {
+			$post   = get_post( $post_id );
+			$blocks = parse_blocks( $post->post_content );
+			foreach ( $blocks as $block ) {
+				if ( 'sensei-lms/featured-video' === $block['blockName'] ) {
+					return trim( render_block( $block ) );
+				}
+			}
+		} else {
+			ob_start();
+			Sensei()->frontend->sensei_lesson_video( $post_id );
+			return trim( ob_get_clean() );
+		}
+	}
```
Really nice, works well as described.

onubrooks reviewed (commented):
Looks good, nothing too serious but left a comment below.

mikeyarce committed:
Revert change to sensei_lesson_video condition

mikeyarce committed:
Render block oembed instead of block HTML

The problem was that just using (render_block) was not creating an iframe, just outputting the URL.  Now we have an iframe outputting.

mikeyarce committed:
Rework get_featured_video_html to support both editors

mikeyarce committed:
PHP Lint fixes

yscik commented on includes/class-sensei-utils.php:
```diff
@@ -2626,6 +2626,51 @@ public static function format_last_activity_date( string $date ) {
+		} else {
+			$video_embed = get_post_meta( $post_id, '_lesson_video_embed', true );
+		}
+		if ( 'http' === substr( $video_embed, 0, 4 ) ) {
```
We should avoiding duplicating code like this. Instead, let's create a new method, for example `render_video_embed`, that can be used in both methods.

mikeyarce replied:
Done! ✅

mikeyarce committed:
Extract out embed render to separate function

mikeyarce committed:
Fix unit test check

yscik reviewed (approved):
Nice work!

yscik commented on includes/class-sensei-utils.php:
```diff
@@ -2626,6 +2626,68 @@ public static function format_last_activity_date( string $date ) {
+				if ( 'sensei-lms/featured-video' === $block['blockName'] ) {
+					// Handle Media Library Files.
+					if ( 'sensei-pro/interactive-video' === $block['innerBlocks'][0]['blockName'] ) {
+						$block = $block['innerBlocks'][0];
```
Oh just started testing with the merged branch and there is an issue here, the interactive video block will get lost because we just take the video URL from it. 

I think we can render the whole featured-video block here and return it, just have to make sure that the right content filters are also applied. I'm investigating a bit but so far looks like just running autembed is enough:

```php
$content = render_block( $block );
global $wp_embed;
return $wp_embed->autoembed( $content );
```
